tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/stratifiedRules/aux/caches/Cache.scala

```diff
@@ -0,0 +1,153 @@
+package at.forsyte.apalache.tla.bmcmt.stratifiedRules.aux.caches
+
+import at.forsyte.apalache.tla.bmcmt.smt.SolverContext
+import at.forsyte.apalache.tla.bmcmt.stratifiedRules.aux.Stackable
+
+import scala.collection.immutable.HashMap
+
+/**
+ * An abstract cache that assigns `TargetT`-typed values to `SourceT` typed values, while potentially modifying a
+ * `ContextT`-typed context.
+ *
+ * Example: IntValueCache assigns [[ArenaCell]]s to integers, while modifying a [[PureArena]].
+ *
+ * While it extends [[Stackable]], there's no need to implement an actual stack data structure. By assigning each cache
+ * entry a level, effectively a timestamp for when it was added, we can proxy the state of the stack, by filtering out
+ * entries, based on their levels.
+ *
+ * @author
+ *   Jure Kukovec
+ */
+abstract class Cache[ContextT, SourceT, TargetT] extends Stackable {
+
+  type LevelT = Int
+
+  /**
+   * A stack level, see [[Stackable]]
+   */
+  private var _level: Int = 0
+
+  // the base map, tracking the level at which each entry was added.
+  private var _cache: Map[SourceT, (TargetT, LevelT)] = HashMap()
+  // inherited classes can read, but not modify without push/pop
+  protected def cache: Map[SourceT, (TargetT, LevelT)] = _cache
+
+  // reverse mapping
+  private var _reverseCache: Map[TargetT, (SourceT, LevelT)] = HashMap()
+  // inherited classes can read, but not modify without push/pop
+  protected def reverseCache: Map[TargetT, (SourceT, LevelT)] = _reverseCache
+
+  def values(): Iterable[TargetT] = _cache.values.map(_._1)
+
+  /**
+   * Create a target value based on the source value and cache it.
+   *
+   * @param context
+   *   the context before creating a new value
+   * @param srcValue
+   *   a source value
+   * @return
+   *   a target value that is going to be cached and the new context
+   */
+  protected def create(context: ContextT, srcValue: SourceT): (ContextT, TargetT)
+
+  /**
+   * Get a previously cached value for a given source value, or return the previously cached one. Whenever a new value
+   * is created, it is cached. The cached value can be later removed by pop.
+   *
+   * @param srcValue
+   *   a source value
+   * @return
+   *   a target value
+   */
+  def getOrCreate(context: ContextT, srcValue: SourceT): (ContextT, TargetT) =
+    if (_cache.contains(srcValue)) {
+      (context, _cache(srcValue)._1)
+    } else {
+      // introduce a new constant
+      val (newContext, targetValue) = create(context, srcValue)
+      addToCache(srcValue, targetValue)
+      (newContext, targetValue)
+    }
+
+  /**
+   * Add a value to the cache.
+   *
+   * @param source
+   *   the source value
+   * @param target
+   *   the cached target value
+   */
+  protected def addToCache(source: SourceT, target: TargetT): Unit = {
+    _cache += source -> (target, level)
+    _reverseCache += target -> (source, level)
+  }
```

Comment:
What is the correct behavior if an equal `source` is added with a different `target` during a "later" snapshot? Should it just overwrite the previously cached value (along with its stored level)?

Reply:
`addToCache` is protected, the interface entry point is `getOrCreate`, which doesn't allow this to happen.

Reply:
Yeah, but any child class could call this directly and overwrite a key, right? Maybe this should be `private`? If else maybe the check for collisions should be moved here?